lib/cloudregister/registerutils.py

```diff
@@ -240,35 +240,77 @@ def fetch_smt_data(cfg, proxies):
-                if srv == region_servers[-1]:
-                    logging.error('None of the servers responded')
-                    logging.error('\tAttempted: %s' % region_servers)
-                    logging.error('Exiting without registration')
-                    sys.exit(1)
-                continue
+        # After the network interface is up, i.e. After=network-online
+        # is satisfied routing on the framework side may not be setup yet
+        # and we may not be able to immediately reach the update
+        # infrastructure.
+        # As time passes is mor likely that the frameowrk completes routing
+        # setup for outbound traffic. Therefore we wait less time the longer
+        # we are in the process of attempting to access the update
+        # infrastrcuture.
+        # Processing delays are handled at two levels; the individual request
+        # timeout and the time we wait between the retry of obtaining the
+        # update server information. For each timeout the starting number
+        # 15 and 20 seconds, respectiveky is an arbitrary choice. The maximum
+        # wait time is dependent on the number of region servers in the
+        # framework. The maximum wait time is calculated as follows
+        #
+        #sumof(num_regions_srvs * 15/i) + sumof(20/j)
+        # where i = 1,2,3
+        #       j = 1,2
+        #
+        # For a configuration with 3 region servers the framework has 112.5
+        # seconds to setup the routing
+        #
+        max_attempts = 3
+        attempt = 0
+        while attempt < max_attempts:
+            retry_cnt = attempt + 1
+            request_timeout = 15/retry_cnt
+            retry_timeout = int(20/retry_cnt)
+            logging.info(
+                'Getting update server information, attempt %d' % retry_cnt
+            )
+            for srv in region_servers:
+                srvName = srv.strip()
+                logging.info('\tUsing region server: %s' % srvName)
+                certFile = cert_dir + '/' + srvName + '.pem'
+                if not os.path.isfile(certFile):
+                    logging.info(
+                        '\tNo cert found: %s skip this server' % certFile
+                    )
+                    continue
+                try:
+                    response = requests.get(
+                        'https://%s/%s' % (srvName, api),
+                        verify=certFile,
+                        timeout=request_timeout,
+                        proxies=proxies
+                    )
+                    if response.status_code == 200:
+                        attempt = max_attempts
+                        break
+                    else:
+                        logging.error('=' * 20)
+                        logging.error(
+                            'Server returned: %d' % response.status_code
+                        )
+                        logging.error('Server error: "%s"' % response.reason)
```

Comment:
`attempts` counter isn't increased in this `else` branch.

If the server returns a non-200 status, but no exception is raised -- wouldn't it be caught in an infinite loop?

There's handling for this down below (`if (not response) or (not response.status_code == 200):`) -- but it's outside the `while attempt` loop.